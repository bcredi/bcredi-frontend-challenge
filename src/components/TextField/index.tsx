import React, { useState } from "react";
import styled from "styled-components";
import { ReactComponent as EyeSvg } from "../../assets/eye-solid.svg";
import { ReactComponent as EyeSlashSvg } from "../../assets/eye-slash-solid.svg";
import { InputWrapper, InputError, Input } from "../../styles";

interface props {
  setEmpty: React.Dispatch<React.SetStateAction<boolean>>;
  error: boolean;
  handleChange: any;
  maxLength?: number;
  minLength?: number;
  value: string;
  name: string;
  placeholder: string;
  validateValueFormat: (value: string) => 1 | any;
}

const TextField = ({
  setEmpty,
  error,
  handleChange,
  maxLength,
  minLength,
  value,
  name,
  placeholder,
  validateValueFormat,
}: props) => {
  const [validValue, setValidValue] = useState(true);
  const [visible, toggleVisible] = useState(false);

  const StyledInputWrapper =
    name === "email"
      ? EmailInputWrapper
      : name === "birthDate" || name === "cpf"
      ? SmallIpuntsWrapper
      : InputWrapper;

  function validateValue(value: string) {
    const valueValidationStatus = validateValueFormat(value);
    if (valueValidationStatus === 1) {
      setEmpty(false);
      return setValidValue(true);
    } else if (valueValidationStatus === `empty ${name}`) {
      setValidValue(true);
    } else {
      setEmpty(false);
      return setValidValue(false);
    }
  }

  function formatValue(Event: React.ChangeEvent<HTMLInputElement>) {
    if (name === "birthDate") {
      Event.target.value = Event.target.value.replace(
        /^(\d{2})(\d{2})(\d{4})$/,
        "$1/$2/$3"
      );
      handleChange(Event);
    } else if (name === "cpf") {
      Event.target.value = Event.target.value.replace(
        /^(\d{3})(\d{3})(\d{3})(\d{2})$/,
        "$1.$2.$3-$4"
      );
      handleChange(Event);
    } else {
      handleChange(Event);
    }
  }

  return (
    <StyledInputWrapper>
      <InputLabel>
        {name === "email"
          ? "E-mail"
          : name === "password"
          ? "Senha"
          : name === "birthDate"
          ? "Data de nascimento"
          : name === "cpf"
          ? "CPF"
          : null}
      </InputLabel>
      <Input
        required
        className="form-input"
        type={name === "password" ? (visible ? "text" : "password") : "text"}
        placeholder={placeholder}
        name={name}
        maxLength={maxLength ? maxLength : 60}
        minLength={minLength && minLength}
        value={value}
        onChange={(e) => {
          formatValue(e);
        }}
        onBlur={(e) => validateValue(e.target.value)}
        onInvalid={() => setEmpty(true)}
        style={{
          border: `1px solid ${
            error || !validValue
              ? "var(--base-color-error)"
              : "var(--base-color-border)"
          }`,
          color: `${
            error || !validValue
              ? "var(--base-color-error"
              : "var(--base-color-text)"
          }`,
        }}
      />
      {name === "password" ? (
        !visible ? (
          <EyeIcon onClick={() => toggleVisible(!visible)} />
        ) : (
          <EyeSlashIcon onClick={() => toggleVisible(!visible)} />
        )
      ) : null}
      <InputError>
        {!validValue
          ? name === "email"
            ? `O e-mail informado não é válido`
            : name === "password"
            ? "A senha precisa ter no mínimo 8 caracteres"
            : name === "birthDate"
            ? "Data inválida"
            : name === "cpf"
            ? "CPF inválido"
            : null
          : error
          ? name === "email"
            ? `O campo e-mail é obrigatório`
            : name === "password"
            ? "O campo senha é obrigatório"
            : name === "birthDate" || name === "cpf"
            ? "O campo é obrigatório"
            : null
          : null}
      </InputError>
    </StyledInputWrapper>
  );
};

const EmailInputWrapper = styled(InputWrapper)`
  margin-top: 0;
  @media (max-width: 600px) {
    margin-top: 0;
  }
`;

const SmallIpuntsWrapper = styled(InputWrapper)`
  width: 148px;
`;

const InputLabel = styled.label`
  font-size: 14px;
  line-height: 18px;
  color: var(--base-color-warm-grey);
  text-align: left;
`;

const EyeIcon = styled(EyeSvg)`
  width: 16px;
  height: 10px;
  position: relative;
  color: var(--base-color-warm-grey);
  margin-left: -30px;
  cursor: pointer;
  &:hover {
    color: var(--base-color-text);
  }
`;

const EyeSlashIcon = styled(EyeSlashSvg)`
  width: 16px;
  height: 10px;
  position: relative;
  color: var(--base-color-warm-grey);
  margin-left: -30px;
  cursor: pointer;
  &:hover {
    color: var(--base-color-text);
  }
`;

export default TextField;
