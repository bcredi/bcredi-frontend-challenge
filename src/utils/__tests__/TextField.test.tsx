import React from "react";
import TextField from "../../components/TextField";
import renderer from "react-test-renderer";
import ReactDOM from "react-dom";

describe("<TextField />", () => {
  // Email Input
  test("Renders email input", () => {
    const wrapper = document.createElement("div");
    ReactDOM.render(
      <TextField
        name="email"
        placeholder="teste@teste.com"
        error={false}
        value={""}
        handleChange={() => {}}
        validateValueFormat={() => {}}
        setEmpty={() => {}}
      />,
      wrapper
    );

    const label = wrapper.querySelector("label");
    expect(label?.firstChild?.textContent).toEqual("E-mail");

    const input = wrapper.querySelector(".form-input");
    expect(input?.getAttribute("type")).toEqual("text");
    expect(input?.getAttribute("name")).toEqual("email");
    expect(input?.getAttribute("placeholder")).toEqual("teste@teste.com");
    expect(input?.getAttribute("maxlength")).toEqual("60");
  });

  // CPF Input
  test("Renders cpf input", () => {
    const wrapper = document.createElement("div");
    ReactDOM.render(
      <TextField
        name="cpf"
        placeholder="000.000.000-00"
        error={false}
        value={""}
        maxLength={14}
        handleChange={() => {}}
        validateValueFormat={() => {}}
        setEmpty={() => {}}
      />,
      wrapper
    );

    const label = wrapper.querySelector("label");
    expect(label?.firstChild?.textContent).toEqual("CPF");

    const input = wrapper.querySelector(".form-input");
    expect(input?.getAttribute("type")).toEqual("text");
    expect(input?.getAttribute("name")).toEqual("cpf");
    expect(input?.getAttribute("placeholder")).toEqual("000.000.000-00");
    expect(input?.getAttribute("maxlength")).toEqual("14");
  });

  // BirthDate Input
  test("Renders birth date input", () => {
    const wrapper = document.createElement("div");
    ReactDOM.render(
      <TextField
        name="birthDate"
        placeholder="DD/MM/AA"
        error={false}
        value={""}
        maxLength={10}
        handleChange={() => {}}
        validateValueFormat={() => {}}
        setEmpty={() => {}}
      />,
      wrapper
    );

    const label = wrapper.querySelector("label");
    expect(label?.firstChild?.textContent).toEqual("Data de nascimento");

    const input = wrapper.querySelector(".form-input");
    expect(input?.getAttribute("type")).toEqual("text");
    expect(input?.getAttribute("name")).toEqual("birthDate");
    expect(input?.getAttribute("placeholder")).toEqual("DD/MM/AA");
    expect(input?.getAttribute("maxlength")).toEqual("10");
  });

  // Password Input
  test("Renders password input", () => {
    const wrapper = document.createElement("div");
    ReactDOM.render(
      <TextField
        name="password"
        placeholder="Cadastre uma senha"
        error={false}
        value={""}
        minLength={8}
        handleChange={() => {}}
        validateValueFormat={() => {}}
        setEmpty={() => {}}
      />,
      wrapper
    );

    const label = wrapper.querySelector("label");
    expect(label?.firstChild?.textContent).toEqual("Senha");

    const input = wrapper.querySelector(".form-input");
    expect(input?.getAttribute("type")).toEqual("password");
    expect(input?.getAttribute("name")).toEqual("password");
    expect(input?.getAttribute("placeholder")).toEqual("Cadastre uma senha");
    expect(input?.getAttribute("minlength")).toEqual("8");
    expect(input?.getAttribute("maxlength")).toEqual("60");
  });

  test("Catches DOM unexpected changes", () => {
    const component = renderer
      .create(
        <TextField
          name="email"
          placeholder="teste@teste.com"
          error={false}
          value={""}
          handleChange={() => {}}
          validateValueFormat={() => {}}
          setEmpty={() => {}}
        />
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });
});
