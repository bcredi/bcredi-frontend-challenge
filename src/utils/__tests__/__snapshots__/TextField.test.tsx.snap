// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> Catches DOM unexpected changes 1`] = `
<div
  className="sc-eCssSg sc-jrAGrp jhDcFB kLrRrm"
>
  <label
    className="sc-iqHYGH fIrSEl"
  >
    E-mail
  </label>
  <input
    className="sc-jSgupP jHzLmv form-input"
    maxLength={60}
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    onInvalid={[Function]}
    placeholder="teste@teste.com"
    required={true}
    style={
      Object {
        "border": "1px solid var(--base-color-border)",
        "color": "var(--base-color-text)",
      }
    }
    type="text"
    value=""
  />
  <div
    className="sc-gKsewC dGAaCy"
  />
</div>
`;

exports[`Catches DOM unexpected changes 1`] = `
<div
  className="sc-eCssSg sc-jrAGrp jhDcFB kLrRrm"
>
  <label
    className="sc-iqHYGH fIrSEl"
  >
    E-mail
  </label>
  <input
    className="sc-jSgupP jHzLmv form-input"
    maxLength={60}
    name="email"
    onBlur={[Function]}
    onChange={[Function]}
    onInvalid={[Function]}
    placeholder="teste@teste.com"
    required={true}
    style={
      Object {
        "border": "1px solid var(--base-color-border)",
        "color": "var(--base-color-text)",
      }
    }
    type="text"
    value=""
  />
  <div
    className="sc-gKsewC dGAaCy"
  />
</div>
`;
